sdk/core/azure-core/src/main/java/com/azure/core/implementation/ByteCountingWritableByteChannel.java

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.util.ProgressReporter;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+/**
+ * Count bytes written to the target channel.
+ */
+public class ByteCountingWritableByteChannel implements WritableByteChannel {
+
+    private final WritableByteChannel channel;
+    private final ProgressReporter progressReporter;
+
+    private static final AtomicLongFieldUpdater<ByteCountingWritableByteChannel> BYTES_WRITTEN_ATOMIC_UPDATER =
+        AtomicLongFieldUpdater.newUpdater(ByteCountingWritableByteChannel.class, "bytesWritten");
+    private volatile long bytesWritten;
+
+    public ByteCountingWritableByteChannel(WritableByteChannel channel, ProgressReporter progressReporter) {
+        this.channel = Objects.requireNonNull(channel, "'channel' must not be null");
+        this.progressReporter = progressReporter;
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+        int written = channel.write(src);
+        BYTES_WRITTEN_ATOMIC_UPDATER.addAndGet(this, written);
+        if (progressReporter != null) {
+            progressReporter.reportProgress(written);
+        }
+        return written;
```

Comment:
Does this need to handle the case where the full `ByteBuffer` isn't written in one write operation?

Reply:
No. That's caller responsibility. We just intercept channel api here, so we're bound to it's contract.